tools/fs/safe-watcher.ts

```diff
@@ -1,476 +1,364 @@
-  // to create this watcher.
-  for (const path of watchRoots) {
-    let relativePath = pathRelative(path, absPath);
-    if (
-      path !== absPath &&
-      !relativePath.startsWith('..') &&
-      !relativePath.startsWith('/')
-    ) {
-      return;
-    }
+  if (watcherEnabled || PRIORITIZE_CHANGED) {
+    // As long as native file watching is enabled (even if it doesn't
+    // work correctly) and the developer hasn't explicitly opted out of
+    // the file watching priority system, poll unchanged files at a
+    // lower frequency (every 5000ms by default).
+    return DEFAULT_POLLING_INTERVAL;
   }
 
-  // TODO: check if there are any existing watchers that are children of this
-  // watcher and stop them
-
-  nsfw(
-    convertToOSPath(absPath),
-    (events) => {
-      events.forEach(event => {
-        if(event.action === nsfw.actions.RENAMED) {
-          let oldPath = nativeJoin(event.directory, event.oldFile);
-          let oldEntry = entries[toPosixPath(oldPath)];
-          if (oldEntry) {
-            oldEntry._fire('rename');
-          }
+  // If native file watching is disabled and the developer has
+  // explicitly opted out of the priority system, poll everything at the
+  // higher frequency (every 500ms by default). Note that this leads to
+  // higher idle CPU usage, so the developer may want to adjust the
+  // METEOR_WATCH_POLLING_INTERVAL_MS environment variable.
+  return NO_WATCHER_POLLING_INTERVAL;
+}
 
-          let path = nativeJoin(event.newDirectory, event.newFile);
-          let newEntry = entries[toPosixPath(path)];
-          if (newEntry) {
-            newEntry._fire('change');
-          }
-        } else {
-            let path = nativeJoin(event.directory, event.file);
-            let entry = entries[toPosixPath(path)];
-            if (entry) {
-              entry._fire(fireNames[event.action]);
-            }
-        }
-      })
+function startPolling(absPath: string, callback: ChangeCallback): SafeWatcher {
+  const osPath = convertToOSPath(absPath);
+  // Initial polling interval.
+  let interval = getPollingInterval(absPath);
+  const pollCallback = (curr: Stats, prev: Stats) => {
+    // Compare modification times to detect a change.
+    if (+curr.mtime !== +prev.mtime) {
+      changedPaths.add(absPath);
+      callback("change");
+    }
+  };
+  watchFile(osPath, { interval }, pollCallback);
+  return {
+    close() {
+      unwatchFile(osPath, pollCallback);
+      changedPaths.delete(absPath);
     }
-  ).then(watcher => {
-    watcher.start()
-  });
+  };
+}
+
+/**
+ * Fall back to polling. If a critical error occurs,
+ * we disable native watching and close all existing native watchers.
+ */
+function fallbackToPolling() {
+  if (watcherEnabled) {
+    console.error("Critical native watcher error encountered. Falling back to polling for all entries.");
+    watcherEnabled = false;
+    closeAllWatchers();
+  }
```

Comment:
I'm keeping the previous Meteor polling strategy as a fallback or configured by the `METEOR_WATCH_FORCE_POLLING` env. ParceLWatcher is supposed to handle native fallbacks across all OS, so this might not be needed anymore. Still, to avoid a breaking change for those using this in certain environments, we're leaving it as is.